app/com/linkedin/drelephant/tuning/engine/SparkHBTParamRecommender.java

```diff
@@ -0,0 +1,326 @@
+    this.lastRunExecutorMemory = lastRunExecutorMemory;
+  }
+
+  public int getLastRunExecutorCore() {
+    return lastRunExecutorCore;
+  }
+
+  public void setLastRunExecutorCore(int lastRunExecutorCore) {
+    this.lastRunExecutorCore = lastRunExecutorCore;
+  }
+
+  public long getDriverMaxPeakJVMUsedMemory() {
+    return driverMaxPeakJVMUsedMemory;
+  }
+
+  public void setDriverMaxPeakJVMUsedMemory(long driverMaxPeakJVMUsedMemory) {
+    this.driverMaxPeakJVMUsedMemory = driverMaxPeakJVMUsedMemory;
+  }
+
+  public long getSuggestedSparkDriverMemory() {
+    return suggestedSparkDriverMemory;
+  }
+
+  public void setSuggestedSparkDriverMemory(long suggestedSparkDriverMemory) {
+    this.suggestedSparkDriverMemory = suggestedSparkDriverMemory;
+  }
+
+  /**
+   * This method returns suggested parameter for the job. Currently it runs following parameters:
+   * executor memory, core, driver memory and memory factor
+   *
+   * @return HashMap with property name as key and double value
+   */
+  public HashMap<String, Double> getHBTSuggestion() {
+    HashMap<String, Double> suggestedParameters = new HashMap<String, Double>();
+    suggestExecutorMemoryCore();
+    suggestMemoryFactor();
+    suggestDriverMemory();
+    suggestedParameters.put(SparkConfigurationConstants.SPARK_EXECUTOR_MEMORY_KEY, (double) suggestedExecutorMemory
+        / FileUtils.ONE_MB);
+    suggestedParameters.put(SparkConfigurationConstants.SPARK_EXECUTOR_CORES_KEY, suggestedCore.doubleValue());
+    if (suggestedMemoryFactor < UnifiedMemoryHeuristic.SPARK_MEMORY_FRACTION_THRESHOLD()) {
+      suggestedMemoryFactor = UnifiedMemoryHeuristic.SPARK_MEMORY_FRACTION_THRESHOLD();
+    }
+    suggestedParameters.put(SparkConfigurationConstants.SPARK_MEMORY_FRACTION_KEY, suggestedMemoryFactor);
+    suggestedParameters.put(SparkConfigurationConstants.SPARK_DRIVER_MEMORY_KEY, (double) suggestedDriverMemory
+        / FileUtils.ONE_MB);
+    logger.info("Following are the suggestions for spark parameters for app id : " + appResult.flowExecId);
+    logger.info("suggestedExecutorMemory " + suggestedExecutorMemory);
+    logger.info("suggestedCore " + suggestedCore);
+    logger.info("suggestedMemoryFactor " + suggestedMemoryFactor);
+    logger.info("suggestedDriverMemory " + suggestedDriverMemory);
+    return suggestedParameters;
+  }
+
+  private Long getMaxPeakJVMUsedMemoryPerCore() {
+    return maxPeakJVMUsedMemory / lastRunExecutorCore;
+  }
+
+  private Long getMaxPeakUnifiedMemoryPerCore() {
+    return maxPeakUnifiedMemory / lastRunExecutorCore;
+  }
+
+  /**
+   * Suggest executor memory and core We calculate per core peak JVM used memory. Then we start from
+   * max executor core and compute memory requirement for max executor core If this is below
+   * threshold of max executor core then good, otherwise reduce the core and check memory
+   * requirement.
+   */
+  private void suggestExecutorMemoryCore() {
```

Comment:
I think , for the next version , we should also think to add number of partitions / tasks parameters  , to come up with final executor memory .

Reply:
Definitely.